gcs/src/main/java/com/google/cloud/hadoop/fs/gcs/RefreshTokenAuth2Provider.java

```diff
@@ -0,0 +1,231 @@
+      String tokenEndpoint,
+      String clientId,
+      RedactedString clientSecret,
+      RedactedString refreshToken,
+      String proxyAddress,
+      RedactedString proxyUsername,
+      RedactedString proxyPassword)
+      throws IOException {
+
+    AccessToken token = null;
+    HttpURLConnection conn = null;
+    logger.atFine().log("Get a new access token using the refresh token grant flow");
+    try {
+      URL url = new URL(tokenEndpoint);
+      conn = (HttpURLConnection) url.openConnection(getProxy(proxyAddress));
+      conn.setRequestMethod("POST");
+      conn.setReadTimeout(READ_TIMEOUT);
+      conn.setConnectTimeout(CONNECT_TIMEOUT);
+      conn.setRequestProperty("Content-Type", "application/x-www-form-urlencoded");
+      conn.setRequestProperty("Accept", "application/json");
+      if (proxyUsername != null && proxyUsername.value() != null) {
+        logger.atFine().log("Setup the request with the proxy auth");
+        conn.setRequestProperty(
+            HttpHeaders.PROXY_AUTHORIZATION,
+            "Basic "
+                + BaseEncoding.base64()
+                    .encode(
+                        (proxyUsername.value() + ":" + proxyPassword.value())
+                            .getBytes(StandardCharsets.UTF_8)));
+      }
+      conn.setDoOutput(true);
+      String payload =
+          "grant_type=refresh_token&refresh_token="
+              + refreshToken.value()
+              + "&client_id="
+              + clientId
+              + "&client_secret="
+              + clientSecret.value();
+      conn.getOutputStream().write(payload.getBytes("UTF-8"));
+
+      int httpResponseCode = conn.getResponseCode();
+      logger.atFine().log("Response " + httpResponseCode);
+
+      String responseContentType = conn.getHeaderField("Content-Type");
+
+      if (httpResponseCode == HttpURLConnection.HTTP_OK
+          && responseContentType.startsWith("application/json")) {
+        logger.atFine().log("Received a 200 with presumably a new access token");
+        InputStream httpResponseStream = conn.getInputStream();
+        token = parseTokenFromStream(httpResponseStream);
+      } else {
+        logger.atFine().log("Refresh token grant flow failed");
+        InputStream stream = conn.getErrorStream();
+        if (stream == null) {
+          // no error stream, try the original input stream
+          stream = conn.getInputStream();
+        }
+        String responseBody = consumeInputStream(stream);
+        logger.atSevere().log(
+            "Received following response '"
+                + responseBody
+                + "' with code '"
+                + httpResponseCode
+                + "'");
+      }
+    } finally {
+      if (conn != null) {
+        conn.disconnect();
+      }
+    }
+    return token;
+  }
+
+  private String consumeInputStream(InputStream stream) throws IOException {
+    BufferedReader br = new BufferedReader(new InputStreamReader(stream, "utf-8"));
```

Comment:
Can we use static constant for "utf-8" ?

Reply:
Wont be application once I use the google client instead.